tests/shouldwork/Cores/Xilinx/DcFifo/Abstract.hs

```diff
@@ -0,0 +1,155 @@
+    readNow = not stall && not fifoEmpty
+{-# NOINLINE fifoSampler #-}
+
+-- | Drives Xilinx FIFO with an ascending sequence of 'BitVector's. Stalls
+-- intermittently based on stall input.
+fifoDriver ::
+  forall n dom depth .
+  ( KnownDomain dom
+  , KnownNat n ) =>
+  Clock dom -> Reset dom -> Enable dom ->
+  -- | Stall circuit? For this test case, this signal comes from 'lfsrF'
+  Signal dom Stall ->
+  -- | Signals from FIFO
+  Signal dom (Full, DataCount depth) ->
+  -- | Maybe write input to FIFO
+  Signal dom (Maybe (BitVector n))
+fifoDriver clk rst ena stalls inps =
+  mealyB clk rst ena go 0 (stalls, inps)
+ where
+  go ::
+    BitVector n ->
+    (Stall, (Full, DataCount depth)) ->
+    (BitVector n, Maybe (BitVector n))
+  go n0 (stall, (full, _dataCount)) = (n1, maybeWrite)
+   where
+    maybeWrite = willWrite `orNothing` n0
+    willWrite = not stall && not full
+    n1 = if willWrite then succ n0 else n0
+
+type ConfiguredFifo read write =
+  Clock write -> Clock read ->
+  -- | Asynchronous reset
+  Reset read ->
+
+  -- | Write data
+  Signal write (Maybe (BitVector 16)) ->
+  -- | Read enable
+  Signal read Bool ->
+  FifoOut read write 4 (BitVector 16)
+
+mkTestBench ::
+  forall read write.
+  (KnownDomain write, KnownDomain read) =>
+  ConfiguredFifo read write ->
+  Signal read Bool
+mkTestBench cFifo = done
+ where
+  (rClk, wClk) = biTbClockGen (not <$> done)
+
+  rRst = resetGen
+  wRst = resetGen
+
+  rEna = enableGen
+  wEna = enableGen
+
+  -- The reset is asynchronous; we make sure the reset is always asserted at
+  -- least one cycle in the slower clock domain.
+  slowReset =
+    if snatToNum @Natural (clockPeriod @write) > snatToNum @Natural (clockPeriod @read)
+      then unsafeToReset (unsafeSynchronizer wClk rClk (unsafeFromReset wRst))
+      else rRst
+
+  -- Driver
+  wLfsr = bitToBool <$> lfsrF wClk wRst wEna 0xDEAD
+  writeData = fifoDriver wClk wRst wEna wLfsr (bundle (isFull, writeCount))
+
+  -- Sampler
+  rLfsr = bitToBool <$> lfsrF rClk rRst rEna 0xBEEF
+  (readEnable, maybeReadData) =
+    unbundle $
+      fifoSampler rClk rRst rEna rLfsr (bundle (isEmpty, readCount, fifoData))
+
+  FifoOut{isFull, writeCount, isEmpty, readCount, fifoData} =
+    cFifo wClk rClk slowReset writeData readEnable
+
+  done = fifoVerifier rClk rRst rEna maybeReadData
+{-# INLINE mkTestBench #-}
+
+fifoVerifier ::
+  KnownDomain dom =>
+  Clock dom -> Reset dom -> Enable dom ->
+  Signal dom (Maybe (BitVector 16)) ->
+  Signal dom Bool
+fifoVerifier clk rst ena actual = done0
+ where
+  expected = regEn clk rst ena 0 (isJust <$> actual) $ expected + 1
+  counter = register clk rst ena (0 :: Signed 32) ((+1) <$> counter)
+  done = (expected .>. 100) .||. (counter .>. 25000)
+  premature = (counter .>. 25000) .&&. (expected .<. 100)
+  expected0 = liftA2 (<$) expected actual
+  done0 =
+    assert clk rst "Doesn't time out" premature (pure False) $
+      assert clk rst "fifoVerifier" actual expected0 done
```

Comment:
Ah, very nice 🙄. Even if you comment on specific code lines in a commit on GitHub, it still formats it as "comment on commit". The above string of comments refers to these lines of code.

Reply:
Yes, you're right. I "tested" it by making the cycle limit small, which does fail, but it fails over many cycles rather than failing and stopping the simulation :slightly_frowning_face: